File: core/src/main/cpp/msg/ResponseMessageBuilder.cpp
/**
 * Copyright (c) 2016-2018 Netflix, Inc.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <msg/ResponseMessageBuilder.h>
#include <MslConstants.h>
#include <MslCryptoException.h>
#include <MslEncodingException.h>
#include <MslEntityAuthException.h>
#include <MslError.h>
#include <MslException.h>
#include <MslInternalException.h>
#include <MslKeyExchangeException.h>
#include <MslMasterTokenException.h>
#include <MslMessageException.h>
#include <MslUserAuthException.h>
#include <crypto/IRandom.h>
#include <crypto/NullCryptoContext.h>
#include <entityauth/EntityAuthenticationData.h>
#include <entityauth/EntityAuthenticationScheme.h>
#include <io/MslEncoderFactory.h>
#include <io/MslEncoderFormat.h>
#include <keyx/KeyExchangeFactory.h>
#include <keyx/KeyRequestData.h>
#include <keyx/KeyResponseData.h>
#include <msg/ErrorHeader.h>
#include <msg/MessageHeader.h>
#include <tokens/MasterToken.h>
#include <tokens/MslUser.h>
#include <tokens/ServiceToken.h>
#include <tokens/TokenFactory.h>
#include <tokens/UserIdToken.h>
#include <userauth/UserAuthenticationData.h>
#include <userauth/UserAuthenticationFactory.h>
#include <userauth/UserAuthenticationScheme.h>
#include <util/MslContext.h>
#include <util/MslStore.h>
#include <util/MslUtils.h>
#include <vector>

using namespace std;
using namespace netflix::msl::crypto;
using namespace netflix::msl::entityauth;
using namespace netflix::msl::io;
using namespace netflix::msl::keyx;
using namespace netflix::msl::tokens;
using namespace netflix::msl::userauth;
using namespace netflix::msl::util;
using namespace netflix::msl::MslConstants;
using KeyExchangeData = netflix::msl::keyx::KeyExchangeFactory::KeyExchangeData;
using HeaderData = netflix::msl::msg::MessageHeader::HeaderData;
using HeaderPeerData = netflix::msl::msg::MessageHeader::HeaderPeerData;

namespace netflix {
namespace msl {
typedef vector<uint8_t> ByteArray;
namespace msg {

namespace {

/**
 * Issue a new master token for the specified identity or renew an existing
 * master token.
 *
 * @param ctx MSL context.
 * @param format MSL encoder format.
 * @param keyRequestData available key request data.
 * @param masterToken master token to renew. Null if the identity is
 *        provided.
 * @param entityAuthData entity authentication data. Null if a master token
 *        is provided.
 * @return the new master token and crypto context or {@code} null if the
 *         factory chooses not to perform key exchange.
 * @throws MslCryptoException if the crypto context cannot be created.
 * @throws MslKeyExchangeException if there is an error with the key
 *         request data or the key response data cannot be created or none
 *         of the key exchange schemes are supported.
 * @throws MslMasterTokenException if the master token is not trusted.
 * @throws MslEncodingException if there is an error parsing or encoding
 *         the JSON.
 * @throws MslEntityAuthException if there is a problem with the master
 *         token identity or entity identity.
 * @throws MslException if there is an error creating or renewing the
 *         master token.
 */
shared_ptr<KeyExchangeData> issueMasterToken(shared_ptr<MslContext> ctx,
		const MslEncoderFormat& format,
		set<shared_ptr<KeyRequestData>> keyRequestData,
		shared_ptr<MasterToken> masterToken,
		shared_ptr<EntityAuthenticationData> entityAuthData)
{
	// Attempt key exchange in the preferred order.
	shared_ptr<IException> keyxException;
	set<shared_ptr<KeyExchangeFactory>> factorySet = ctx->getKeyExchangeFactories();
	for (set<shared_ptr<KeyExchangeFactory>>::iterator factories = factorySet.begin();
		 factories != factorySet.end();
		 ++factories)
	{
		shared_ptr<KeyExchangeFactory> factory = *factories;
		set<shared_ptr<KeyRequestData>>::iterator requests = keyRequestData.begin();
		while (requests != keyRequestData.end()) {
			shared_ptr<KeyRequestData> request = *requests;
			if (factory->getScheme() != request->getKeyExchangeScheme()) {
			    requests++;
				continue;
			}

			// Attempt the key exchange, but if it fails try with the next
			// combination before giving up.
			try {
				if (masterToken)
					return factory->generateResponse(ctx, format, request, masterToken);
				else
					return factory->generateResponse(ctx, format, request, entityAuthData);
			} catch (const MslCryptoException& e) {
				if (requests == keyRequestData.end()) throw e;
				keyxException = e.clone();
			} catch (const MslKeyExchangeException& e) {
				if (requests == keyRequestData.end()) throw e;
				keyxException = e.clone();
			} catch (const MslEncodingException& e) {
				if (requests == keyRequestData.end()) throw e;
				keyxException = e.clone();
			} catch (const MslMasterTokenException& e) {
				if (requests == keyRequestData.end()) throw e;
				keyxException = e.clone();
			} catch (const MslEntityAuthException& e) {
				if (requests == keyRequestData.end()) throw e;
				keyxException = e.clone();
			}

			requests++;
		}
	}

	// We did not perform a successful key exchange. If we caught an
	// exception then throw that exception now.
	if (keyxException) {
		MslUtils::rethrow(keyxException);
		throw MslInternalException("Unexpected exception caught during key exchange.", *keyxException);
	}

	// If we didn't find any then we're unable to perform key exchange.
	stringstream ss;
	ss << "[ ";
	for (set<shared_ptr<KeyRequestData>>::iterator keyRequests = keyRequestData.begin();
		 keyRequests != keyRequestData.end();
		 ++keyRequests)
	{
		ss << *keyRequests << " ";
	}
	ss << "]";
	throw MslKeyExchangeException(MslError::KEYX_FACTORY_NOT_FOUND, ss.str());
}

} // namespace anonymous


ResponseMessageBuilder::ResponseMessageBuilder(
		shared_ptr<MslContext> ctx,
        shared_ptr<MessageHeader> requestHeader)
    : MessageBuilder(ctx)
{
	shared_ptr<MasterToken> masterToken = requestHeader->getMasterToken();
	shared_ptr<EntityAuthenticationData> entityAuthData = requestHeader->getEntityAuthenticationData();
	shared_ptr<UserIdToken> userIdToken = requestHeader->getUserIdToken();
	shared_ptr<UserAuthenticationData> userAuthData = requestHeader->getUserAuthenticationData();

	// The response message ID must be equal to the request message ID + 1.
	const int64_t requestMessageId = requestHeader->getMessageId();
	const int64_t messageId = incrementMessageId(requestMessageId);

	// Compute the intersection of the request and response message
	// capabilities.
	shared_ptr<MessageCapabilities> capabilities = MessageCapabilities::intersection(requestHeader->getMessageCapabilities(), ctx->getMessageCapabilities());

	// Identify the response format.
	shared_ptr<MslEncoderFactory> encoder = ctx->getMslEncoderFactory();
	set<MslEncoderFormat> formats;
	if (capabilities) formats = capabilities->getEncoderFormats();
	const MslEncoderFormat format = encoder->getPreferredFormat(formats);

	try {
		// If the message contains key request data and is renewable...
		shared_ptr<KeyExchangeData> keyExchangeData;
		set<shared_ptr<KeyRequestData>> keyRequestData = requestHeader->getKeyRequestData();
		if (requestHeader->isRenewable() && !keyRequestData.empty()) {
			// If the message contains a master token...
			if (masterToken) {
				// If the master token is renewable or expired then renew
				// the master token.
				if (masterToken->isRenewable() || masterToken->isExpired())
					keyExchangeData = issueMasterToken(ctx, format, keyRequestData, masterToken, shared_ptr<EntityAuthenticationData>());
				// Otherwise we don't need to do anything special.
				else
					keyExchangeData.reset();
			}

			// Otherwise use the entity authentication data to issue a
			// master token.
			else {
				// The message header is already authenticated via the
				// entity authentication data's crypto context so we can
				// simply proceed with the master token issuance.
				keyExchangeData = issueMasterToken(ctx, format, keyRequestData, shared_ptr<MasterToken>(), entityAuthData);
			}
		}

		// If the message does not contain key request data there is no key
		// exchange for us to do.
		else {
			keyExchangeData.reset();
		}

		// If we successfully performed key exchange, use the new master
		// token for user authentication.
		shared_ptr<MasterToken> userAuthMasterToken;
		if (keyExchangeData) {
			userAuthMasterToken = keyExchangeData->keyResponseData->getMasterToken();
		} else {
			userAuthMasterToken = masterToken;
		}

		// If the message contains a user ID token issued by the local
		// entity...
		if (userIdToken && userIdToken->isVerified()) {
			// If the user ID token is renewable and the message is
			// renewable, or it is expired, or it needs to be rebound
			// to the new master token then renew the user ID token.
			if ((userIdToken->isRenewable() && requestHeader->isRenewable()) ||
					userIdToken->isExpired() ||
					!userIdToken->isBoundTo(userAuthMasterToken))
			{
				shared_ptr<TokenFactory> tokenFactory = ctx->getTokenFactory();
				userIdToken = tokenFactory->renewUserIdToken(ctx, userIdToken, userAuthMasterToken);
			}
		}

		// If the message is renewable and contains user authentication
		// data and a master token then we need to attempt user
		// authentication and issue a user ID token.
		else if (requestHeader->isRenewable() && userAuthMasterToken && userAuthData) {
			// If this request was parsed then its user authentication data
			// should have been authenticated and the user will exist. If
			// it was not parsed, then we need to perform user
			// authentication now.
			shared_ptr<MslUser> user = requestHeader->getUser();
			if (!user) {
				const UserAuthenticationScheme scheme = userAuthData->getScheme();
				shared_ptr<UserAuthenticationFactory> factory = ctx->getUserAuthenticationFactory(scheme);
				if (!factory) {
					throw MslUserAuthException(MslError::USERAUTH_FACTORY_NOT_FOUND, scheme.name())
					.setMasterToken(masterToken)
					.setUserAuthenticationData(userAuthData)
					.setMessageId(requestMessageId);
				}
				user = factory->authenticate(ctx, userAuthMasterToken->getIdentity(), userAuthData, shared_ptr<UserIdToken>());
			}
			shared_ptr<TokenFactory> tokenFactory = ctx->getTokenFactory();
			userIdToken = tokenFactory->createUserIdToken(ctx, user, userAuthMasterToken);
		}

		// Create the message builder.
		//
		// Peer-to-peer responses swap the tokens.
		shared_ptr<KeyResponseData> keyResponseData = requestHeader->getKeyResponseData();
		set<shared_ptr<ServiceToken>> serviceTokens = requestHeader->getServiceTokens();
		if (ctx->isPeerToPeer()) {
			shared_ptr<MasterToken> peerMasterToken = (keyResponseData) ? keyResponseData->getMasterToken() : requestHeader->getPeerMasterToken();
			shared_ptr<UserIdToken> peerUserIdToken = requestHeader->getPeerUserIdToken();
			set<shared_ptr<ServiceToken>> peerServiceTokens = requestHeader->getPeerServiceTokens();
			initializeMessageBuilder(messageId, capabilities, peerMasterToken, peerUserIdToken, peerServiceTokens, masterToken, userIdToken, serviceTokens, keyExchangeData);
		} else {
			shared_ptr<MasterToken> localMasterToken = (keyResponseData) ? keyResponseData->getMasterToken() : masterToken;
			initializeMessageBuilder(messageId, capabilities, localMasterToken, userIdToken, serviceTokens, shared_ptr<MasterToken>(), shared_ptr<UserIdToken>(), set<shared_ptr<ServiceToken>>(), keyExchangeData);
		}
	} catch (MslException& e) {
		e.setMasterToken(masterToken);
		e.setEntityAuthenticationData(entityAuthData);
		e.setUserIdToken(userIdToken);
		e.setUserAuthenticationData(userAuthData);
		e.setMessageId(requestMessageId);
		MslUtils::rethrow(e);
		throw e;
	}
}

}}} // namespace netflix::msl::msg

File: core/src/main/javascript/msg/ResponseMessageBuilder.js
/**
 * Copyright (c) 2012-2018 Netflix, Inc.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * <p>A message builder provides methods for building messages.</p>
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
    "use strict";

    var MslConstants = require('../MslConstants.js');
    var MslInternalException = require('../MslInternalException.js');
    var AsyncExecutor = require('../util/AsyncExecutor.js');
    var MslKeyExchangeException = require('../MslKeyExchangeException.js');
    var MslError = require('../MslError.js');
    var MslException = require('../MslException.js');
    var MslUserAuthException = require('../MslUserAuthException.js');
    var MessageCapabilities = require('../msg/MessageCapabilities.js');
    var ErrorHeader = require('../msg/ErrorHeader.js');
    var Class = require('../util/Class.js');
    var MslMessageException = require('../MslMessageException.js');
    var MessageBuilder = require('../msg/MessageBuilder.js');
    var MessageHeader = require('../msg/MessageHeader.js');
    var ServiceToken = require('../tokens/ServiceToken.js');
    var NullCryptoContext = require('../crypto/NullCryptoContext.js');
    var MslUtils = require('../util/MslUtils.js');

    /**
     * Issue a new master token for the specified identity or renew an existing
     * master token.
     *
     * @param {MslContext} ctx MSL context.
     * @param {MslEncoderFormat} format MSL encoder format.
     * @param {Array.<KeyRequestData>} keyRequestData available key request data.
     * @param {MasterToken} masterToken master token to renew. Null if the identity is
     *        provided.
     * @param {EntityAuthenticationData} entityAuthData entity authentication data. Null if a master token
     *        is provided.
     * @param {{result: function(KeyExchangeData), error: function(Error)}}
     *        callback the callback that will receive the key exchange data, or
     *        null if the factory chooses not to perform key exchange, or any
     *        thrown exceptions.
     * @throws MslCryptoException if the crypto context cannot be created.
     * @throws MslKeyExchangeException if there is an error with the key
     *         request data or the key response data cannot be created or none
     *         of the key exchange schemes are supported.
     * @throws MslMasterTokenException if the master token is not trusted.
     * @throws MslEncodingException if there is an error parsing or encoding
     *         the JSON.
     * @throws MslEntityAuthException if there is a problem with the master
     *         token identity or entity identity.
     * @throws MslException if there is an error creating or renewing the
     *         master token.
     */
    function issueMasterToken(ctx, format, keyRequestData, masterToken, entityAuthData, callback) {
        var factoryIndex = 0, requestIndex = 0;
        var factories = ctx.getKeyExchangeFactories();
        var keyxException;
        var entityToken = (masterToken) ? masterToken : entityAuthData;

        // Attempt key exchange in the preferred order.
        function nextExchange() {
            AsyncExecutor(callback, function() {
                // If we've reached the end of the key request data, try them all
                // again with the next factory.
                if (requestIndex >= keyRequestData.length) {
                    requestIndex = 0;
                    ++factoryIndex;
                }

                // If we've reached the end of the factories then stop.
                if (factoryIndex >= factories.length) {
                    // We did not perform a successful key exchange. If we caught an
                    // exception then throw that exception now.
                    if (keyxException)
                        throw keyxException;

                    // If we didn't find any then we're unable to perform key exchange.
                    throw new MslKeyExchangeException(MslError.KEYX_FACTORY_NOT_FOUND, keyRequestData);
                }

                // Grab this iteration's factory and request.
                var factory = factories[factoryIndex];
                var request = keyRequestData[requestIndex];
                if (factory.scheme != request.keyExchangeScheme) {
                    // Try the next request.
                    ++requestIndex;
                    nextExchange();
                    return;
                }

                // Attempt the key exchange.
                factory.generateResponse(ctx, format, request, entityToken, {
                    result: function(keyExchangeData) {
                        // Deliver the result.
                        callback.result(keyExchangeData);
                    },
                    error: function(e) {
                        AsyncExecutor(callback, function() {
                            // Immediately deliver anything that's not a MslException.
                            if (!(e instanceof MslException))
                                throw e;

                            // Otherwise save this exception and try the next
                            // combination.
                            keyxException = e;
                            ++requestIndex;
                            nextExchange();
                        });
                    }
                });
            });
        }

        nextExchange();
    };

    /**
     * Performs key exchange for the request message header if key exchange
     * should occur.
     *
     * @param {MslContext} ctx MSL context.
     * @param {MslEncoderFormat} format MSL encoder format.
     * @param {MessageHeader} requestHeader message with which to attempt key exchange.
     * @param {?EntityAuthenticationData} entityAuthData message header entity authentication data.
     * @param {?MasterToken} message header master token.
     * @param {{result: function(KeyExchangeData), error: function(Error)}}
     *        callback the callback that will receive the key exchange data or
     *        any thrown exceptions.
     * @throws MslCryptoException if the crypto context cannot be created.
     * @throws MslKeyExchangeException if there is an error with the key
     *         request data or the key response data cannot be created or none
     *         of the key exchange schemes are supported.
     * @throws MslMasterTokenException if the master token is not trusted.
     * @throws MslEncodingException if there is an error parsing or encoding
     *         the JSON.
     * @throws MslEntityAuthException if there is a problem with the master
     *         token identity or entity identity.
     * @throws MslException if there is an error creating or renewing the
     *         master token.
     */
    function performKeyExchange(ctx, format, requestHeader, entityAuthData, masterToken, callback) {
        AsyncExecutor(callback, function() {
            // If the message contains key request data and is renewable...
            var keyRequestData = requestHeader.keyRequestData;
            if (requestHeader.isRenewable() && keyRequestData.length > 0) {
                // If the message contains a master token...
                if (masterToken) {
                    // If the master token is renewable or expired then renew
                    // the master token.
                    if (masterToken.isRenewable(null) || masterToken.isExpired(null)) {
                        issueMasterToken(ctx, format, keyRequestData, masterToken, null, callback);
                    } else {
                        return null;
                    }
                }

                // Otherwise use the entity authentication data to issue a
                // master token.
                else {
                    // The message header is already authenticated via the
                    // entity authentication data's crypto context so we can
                    // simply proceed with the master token issuance.
                    issueMasterToken(ctx, format, keyRequestData, null, entityAuthData, callback);
                }
            }

            // If the message does not contain key request data there is no key
            // exchange for us to do.
            else {
                return null;
            }
        });
    }

    /**
     * Return the user ID token that should be included in the message header
     * creating in response to a request. The request's user ID token may be
     * renewed or a new user ID token issued if user authentication data is
     * provided.
     *
     * @param {MslContext} ctx MSL context.
     * @param {MessageHeader} requestHeader request message header.
     * @param {MasterToken} masterToken master token to verify/bind the user ID
     *        token against.
     * @param {{result: function(UserIdToken), error: function(Error)}} callback
     *        the callback that will receive the user ID token or any thrown
     *        exceptions.
     * @throws MslUserAuthException if the user authentication scheme is not
     *         supported.
     */
    function getUserIdToken(ctx, requestHeader, masterToken, callback) {
        AsyncExecutor(callback, function() {
            var userIdToken = requestHeader.userIdToken;
            var userAuthData = requestHeader.userAuthenticationData;
            var requestMessageId = requestHeader.messageId;
            var tokenFactory = ctx.getTokenFactory();

            // If the message contains a user ID token issued by the local
            // entity...
            if (userIdToken && userIdToken.isVerified()) {
                // If the user ID token is renewable and the message is
                // renewable, or it is expired, or it needs to be rebound
                // to the new master token then renew the user ID token.
                if ((userIdToken.isRenewable(null) && requestHeader.isRenewable()) ||
                    userIdToken.isExpired(null) ||
                    !userIdToken.isBoundTo(masterToken))
                {
                    tokenFactory.renewUserIdToken(ctx, userIdToken, masterToken, callback);
                    return;
                }
            }

            // If the message is renewable and contains user authentication
            // data and a master token then we need to attempt user
            // authentication and issue a user ID token.
            else if (requestHeader.isRenewable() && masterToken && userAuthData) {
                // If this request was parsed then its user authentication data
                // should have been authenticated and the user will exist. If
                // it was not parsed, then we need to perform user
                // authentication now.
                var user = requestHeader.user;
                if (!user) {
                    var scheme = userAuthData.scheme;
                    var factory = ctx.getUserAuthenticationFactory(scheme);
                    if (!factory) {
                        throw new MslUserAuthException(MslError.USERAUTH_FACTORY_NOT_FOUND, scheme)
                            .setMasterToken(masterToken)
                            .setUserAuthenticationData(userAuthData)
                            .setMessageId(requestMessageId);
                    }
                    factory.authenticate(ctx, masterToken.identity, userAuthData, null, {
                        result: function(user) {
                            tokenFactory.createUserIdToken(ctx, user, masterToken, callback);
                        },
                        error: callback.error,
                    });
                } else {
                    tokenFactory.createUserIdToken(ctx, user, masterToken, callback);
                }
                return;
            }

            // Otherwise return the header's user ID token (may be null).
            return userIdToken;
        });
    }

    var ResponseMessageBuilder = module.exports = MessageBuilder.extend({

        /**
         * Create a new message builder that will craft a new message in response
         * to another message. The constructed message may be used as a request.
         *
         * @param {MslContext} ctx MSL context.
         * @param {MessageHeader} requestHeader message header to respond to.
         * @param {{result: function(ResponseMessageBuilder), error: function(Error)}}
         *        callback the callback that will receive the message builder or
         *        any thrown exceptions.
         * @throws MslMasterTokenException if the provided message's master token
         *         is not trusted.
         * @throws MslCryptoException if the crypto context from a key exchange
         *         cannot be created.
         * @throws MslKeyExchangeException if there is an error with the key
         *         request data or the key response data cannot be created.
         * @throws MslUserAuthException if there is an error with the user
         *         authentication data or the user ID token cannot be created.
         * @throws MslException if a user ID token in the message header is not
         *         bound to its corresponding master token or there is an error
         *         creating or renewing the master token.
         */
        init: function init(ctx, requestHeader, callback) {
            var self = this;
            AsyncExecutor(callback, function() {
                var masterToken = requestHeader.masterToken;
                var entityAuthData = requestHeader.entityAuthenticationData;
                var userIdToken = requestHeader.userIdToken;
                var userAuthData = requestHeader.userAuthenticationData;

                // The response message ID must be equal to the request message ID + 1.
                var requestMessageId = requestHeader.messageId;
                var messageId = MessageBuilder.incrementMessageId(requestMessageId);

                // Compute the intersection of the request and response message
                // capabilities.
                var capabilities = MessageCapabilities.intersection(requestHeader.messageCapabilities, ctx.getMessageCapabilities());

                // Identify the response format.
                var encoder = ctx.getMslEncoderFactory();
                var formats = (capabilities) ? capabilities.encoderFormats : null;
                var format = encoder.getPreferredFormat(formats);

                // Perform key exchange.
                performKeyExchange(ctx, format, requestHeader, entityAuthData, masterToken, {
                    result: function(keyExchangeData) {
                        AsyncExecutor(callback, function() {
                            // If we successfully performed key exchange, use the new master
                            // token for user authentication.
                            var userAuthMasterToken = (keyExchangeData)
                                ? keyExchangeData.keyResponseData.masterToken
                                : userAuthMasterToken = masterToken;

                            // Grab the local entity authentication data.
                            ctx.getEntityAuthenticationData(null, {
                                result: function(entityAuthData) {
                                    AsyncExecutor(callback, function() {
                                        // Grab the user ID token for this response.
                                        getUserIdToken(ctx, requestHeader, userAuthMasterToken, {
                                            result: function(token) {
                                                AsyncExecutor(callback, function() {
                                                    userIdToken = token;

                                                    // Create the message builder.
                                                    //
                                                    // Peer-to-peer responses swap the tokens.
                                                    var keyResponseData = requestHeader.keyResponseData;
                                                    var serviceTokens = requestHeader.serviceTokens;
                                                    if (ctx.isPeerToPeer()) {
                                                        var peerMasterToken = (keyResponseData) ? keyResponseData.masterToken : requestHeader.peerMasterToken;
                                                        var peerUserIdToken = requestHeader.peerUserIdToken;
                                                        var peerServiceTokens = requestHeader.peerServiceTokens;
                                                        this.initializeMessageBuilder(ctx, messageId, capabilities, entityAuthData, peerMasterToken, peerUserIdToken, peerServiceTokens, masterToken, userIdToken, serviceTokens, keyExchangeData);
                                                        return this;
                                                    } else {
                                                        var localMasterToken = (keyResponseData) ? keyResponseData.masterToken : masterToken;
                                                        this.initializeMessageBuilder(ctx, messageId, capabilities, entityAuthData, localMasterToken, userIdToken, serviceTokens, null, null, null, keyExchangeData);
                                                        return this;
                                                    }
                                                }, self);
                                            },
                                            error: handleError,
                                        });
                                    }, self);
                                },
                                error: handleError,
                            });
                        }, self);
                    },
                    error: handleError,
                });

                function handleError(e) {
                    AsyncExecutor(callback, function() {
                        if (e instanceof MslException) {
                            e.setMasterToken(masterToken);
                            e.setEntityAuthenticationData(entityAuthData);
                            e.setUserIdToken(userIdToken);
                            e.setUserAuthenticationData(userAuthData);
                            e.setMessageId(requestMessageId);
                        }
                        throw e;
                    }, self);
                }
            }, self);
        }
    });

    /**
     * Create a new message builder that will craft a new message in response
     * to another message. The constructed message may be used as a request.
     *
     * @param {MslContext} ctx MSL context.
     * @param {MessageHeader} requestHeader message header to respond to.
     * @param {{result: function(ResponseMessageBuilder), error: function(Error)}}
     *        callback the callback that will receive the message builder or
     *        any thrown exceptions.
     * @throws MslMasterTokenException if the provided message's master token
     *         is not trusted.
     * @throws MslCryptoException if the crypto context from a key exchange
     *         cannot be created.
     * @throws MslKeyExchangeException if there is an error with the key
     *         request data or the key response data cannot be created.
     * @throws MslUserAuthException if there is an error with the user
     *         authentication data or the user ID token cannot be created.
     * @throws MslException if a user ID token in the message header is not
     *         bound to its corresponding master token or there is an error
     *         creating or renewing the master token.
     */
    var ResponseMessageBuilder$create = function ResponseMessageBuilder$create(ctx, requestHeader, callback) {
        new ResponseMessageBuilder(ctx, requestHeader, callback);
    };

    // Exports.
    module.exports.create = ResponseMessageBuilder$create;
})(require, (typeof module !== 'undefined') ? module : mkmodule('ResponseMessageBuilder'));
